generators/commons/src/ast/AbstractWriter.ts

```diff
@@ -0,0 +1,161 @@
+import { AbstractAstNode } from "./AbstractAstNode";
+import { CodeBlock } from "./CodeBlock";
+
+const TAB_SIZE = 4;
+
+export class AbstractWriter {
+    /* The contents being written */
+    public buffer = "";
+    /* Indentation level (multiple of 4) */
+    private indentLevel = 0;
+    /* Whether anything has been written to the buffer */
+    private hasWrittenAnything = false;
+    /* Whether the last character written was a newline */
+    private lastCharacterIsNewline = false;
+
+    /**
+     * Writes arbitrary text
+     * @param text
+     */
+    public write(text: string): void {
+        const textEndsInNewline = text.length > 0 && text.endsWith("\n");
+        // temporarily remove the trailing newline, since we don't want to add the indent prefix after it
+        const textWithoutNewline = textEndsInNewline ? text.substring(0, text.length - 1) : text;
+
+        const indent = this.getIndentString();
+        let indentedText = textWithoutNewline.replaceAll("\n", `\n${indent}`);
+        if (this.isAtStartOfLine()) {
+            indentedText = indent + indentedText;
+        }
+        if (textEndsInNewline) {
+            indentedText += "\n";
+        }
+        this.writeInternal(indentedText);
+    }
+
+    /**
+     * Writes a node
+     * @param node
+     */
+    public writeNode(node: AbstractAstNode): void {
+        node.write(this);
+    }
+
+    /**
+     * Writes a node but then suffixes with a `;` and new line
+     * @param node
+     */
+    public writeNodeStatement(node: AbstractAstNode): void {
+        node.write(this);
+        this.write(";");
+        this.writeNewLineIfLastLineNot();
+    }
+
+    /**
+     * Writes text but then suffixes with a `;`
+     * @param node
+     */
+    public writeTextStatement(text: string): void {
+        const codeBlock = new CodeBlock(text);
+        codeBlock.write(this);
+        this.write(";");
+        this.writeNewLineIfLastLineNot();
+    }
+
+    /**
+     * Writes text but then suffixes with a `;`
+     * @param node
+     */
+    public controlFlow(prefix: string, statement: AbstractAstNode): void {
```

Comment:
Similar to above - control flow and blocks will differ based on the language, so the decision applies here too.